Add documentation (#35)
jolicode/castor: doc/04-context.md

## Context

> For every command that castor execute, it uses a context object. This object contains the default value for the `exec` 
> function (directory, environment variables, pty, tty, etc...).

Reviewer: ```suggestion
For every command that castor execute, it uses a context object. This object contains the default value for the `exec`  or `watch`
functions (directory, environment variables, pty, tty, etc...).
```

Reviewer 2: I'm not sure we should mention the watch function here, it's usage should be quite rare

It also contains custom values that can be set by the user and reused in commands.

The context is immutable, which means that every time you change a value, a new context is created.

### Using the context

> The context is passed to the function if it has a `Castor\Context` type hint.

Reviewer 2: ```suggestion
The context is passed to the function if it has an argument type-hinted with `Castor\Context`.
```

```php
use Castor\Context;

#[AskTask]
function foo(Context $context): void
{
    echo $context->currentDirectory; // will print the directory of the .castor.php file

>     $context = $context->withCd('/tmp'); // will create a new context with the current directory set to /tmp

Reviewer 2: ```suggestion
    $context = $context->withPath('/tmp'); // will create a new context with the current directory set to /tmp
```

    exec('pwd', context: $context); // will print /tmp
}
```

### Creating a new context

> You can create a new context by declaring a function with the `Castor\Attribute\AsContext` attribute.

Reviewer: ```suggestion
You can create a new context by declaring a function with the `Castor\Attribute\AsContext` attribute:
```

```php
use Castor\Attribute\AsContext;
use Castor\Context;

#[AsContext]
function my_context(): Context
{
    return new Context(environment: ['FOO' => 'BAR']);
}

> #[AskTask]

Reviewer 2: ```suggestion
#[AsTask]
```
